odoo/tools/profiler.py

```diff
@@ -242,13 +242,53 @@ def post_process(self):
         super().post_process()
 
 
+class QwebCollector(Collector):
+    """
+    Record qweb execution with directive trace.
+    """
+    name = 'qweb'
+
+    def __init__(self):
+        super().__init__()
+        def hook(results):
+            """
+            param data: {
+                    'archs': {view_id: string},
+                    'data': [{
+                        'view_id': number
+                        'arch': html
+                        'xpath': string
+                        'directive': string
+                        'time': float
+                        'delay': float
+                        'query': number
+                    }]
+                }
+            """
+            self.add({'results': results})
+        self.hook = hook
+
+    def start(self):
+        init_thread = self.profiler.init_thread
+        if not hasattr(init_thread, 'qweb_hooks'):
+            init_thread.qweb_hooks = []
+        if self.hook in init_thread.qweb_hooks:
+            _logger.warning('record qweb recursive call')
+        else:
+            init_thread.qweb_hooks.append(self.hook)
+
+    def stop(self):
+        self.profiler.init_thread.qweb_hooks.remove(self.hook)
+
+
 class ExecutionContext:
     """
     Add some context on thread at current call stack level.
     This context stored by collector beside stack and is used by Speedscope
     to add a level to the stack with this information.
     """
     def __init__(self, **context):
+        self.adjust_level = context.pop('adjust_level', 0)
```

Comment:
You could make adjust_level an explicit parameter. (But I'm not sure it is a good idea to play like that with the level)

It could also be renamed to something like level_offset and maybe be added to the stack_trace_level